File: interview/algorithms/heap_sort.py
import os
import sys


def parent(i):
    return int((i-1)/2)

def left(i):
    return 2*i + 1

def right(i):
    return 2*i + 2

def max_heapify(arr, index, heap_size):
    left_index = left(index)
    right_index = right(index)
    largest = index
    if heap_size > left_index and arr[left_index] > arr[index]:
        largest = left_index
    if heap_size > right_index and arr[right_index] > arr[largest]:
        largest = right_index

    if largest != index:
        tmp = arr[largest]
        arr[largest] = arr[index]
        arr[index] = tmp
        max_heapify(arr, largest, heap_size)


def build_max_heap(arr):
    heap_size = len(arr)
    index = int(len(arr)/2)-1
    while True:
        if index < 0:
            break
        else:
            max_heapify(arr, index, heap_size)
            index = index - 1
  

def heap_sort(arr):
    build_max_heap(arr)
    heap_size = len(arr)
    for i in range(len(arr)-1, 0, -1):
        tmp = arr[i]
        arr[i] = arr[0]
        arr[0] = tmp
        heap_size = heap_size - 1
        max_heapify(arr, 0, heap_size)


if __name__ == "__main__":
    # prepare input
    arr = [4, 1, 3, 2, 16, 9, 10, 14, 8, 7]
    #arr = [5, 13, 2, 25, 7, 17, 20, 8, 4]

    print('un-sorted array {}'.format(arr))
    heap_sort(arr)
    print('sorted array {}'.format(arr))

File: interview/algorithms/sortings/heap_sort.cpp
// heap sorting

#include <iostream>
#include <algorithm>
#include <random>
#include <iterator>
#include <limits>
using namespace std;

#define LEFT(x) (2*x + 1)
#define RIGHT(x) (2*x + 2)
#define PARENT(x) ((x-1)/2)

int HEAP_SIZE{};

template <typename T>
void max_heapify(T arr[], size_t index) {
    size_t l_idx = LEFT(index);
    size_t r_idx = RIGHT(index);
    size_t largest_idx = index;
    if(l_idx < HEAP_SIZE && arr[index] < arr[l_idx]) {
        largest_idx = l_idx;
    }
    if (r_idx < HEAP_SIZE && arr[largest_idx] < arr[r_idx]) {
        largest_idx = r_idx;
    }

    if(index != largest_idx) {
        std::swap(arr[index], arr[largest_idx]);
        max_heapify(arr, largest_idx);
    }
}

template <typename T, size_t N>
void build_max_heap(T (&arr)[N]) {
    HEAP_SIZE = N;
    for(int i = N/2 - 1; i >=0; --i) {
        max_heapify(arr, i);
    }    
}

template <typename T, size_t N>
void heap_sort(T (&arr)[N]) {
    build_max_heap(arr);
    for(int i = N-1; i >= 1; --i) {
        std::swap(arr[0], arr[i]);
        --HEAP_SIZE;
        max_heapify(arr, 0);
    }
}

int main() {
    std::random_device r;
    std::default_random_engine engine(r());
    std::uniform_int_distribution<int> dist(-200, 200);
    auto gen = [dist=std::move(dist), engine=std::move(engine)] () mutable  {
        return dist(engine);
    };
    int arr[10] {};
    std::generate(begin(arr), end(arr), gen);

    cout << "Elements before\n";
    std::copy(begin(arr), end(arr), ostream_iterator<int>(cout, " "));

    cout << "\n\nElements after\n";
    heap_sort(arr);
    std::copy(begin(arr), end(arr), ostream_iterator<int>(cout, " "));
    
}